File: modules/tools/map_datachecker/static_align.py
#!/usr/bin/env python

###############################################################################
# Copyright 2019 The Apollo Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
###############################################################################

import os
import sys
import logging
import yaml
import math
import time
import grpc
script_path = os.path.dirname(os.path.realpath(__file__))
sys.path.append(os.path.join(script_path, 'py_proto'))
import collection_service_pb2_grpc
import collection_service_pb2
import collection_check_message_pb2
import collection_error_code_pb2 as ErrorCode

class StaticAlign:
    def __init__(self, conf_file):
        self._conf_file = conf_file
        with open(conf_file, 'r') as fp:
            self._conf = yaml.load(fp.read())["static_align"]
            self._host = self._conf['grpc_host']
            self._port = self._conf['grpc_port']
        channel = grpc.insecure_channel(self._host + ':' + self._port)
        self.stub = collection_service_pb2_grpc.CollectionCheckerServiceStub(channel)

    def _exception_handler(self, error_code):
        if error_code == ErrorCode.SUCCESS:
            logging.info("ErrorCode.SUCCESS, ignore")
        elif error_code == ErrorCode.ERROR_REPEATED_START:
            logging.warn("ErrorCode.ERROR_REPEATED_START, ignore")
        elif error_code == ErrorCode.ERROR_CHECK_BEFORE_START:
            logging.warn("ErrorCode.ERROR_CHECK_BEFORE_START")
        elif error_code == ErrorCode.ERROR_REQUEST:
            logging.warn("ErrorCode.ERROR_REQUEST")
        elif error_code == ErrorCode.ERROR_GNSS_SIGNAL_FAIL:
            logging.error("ErrorCode.ERROR_GNSS_SIGNAL_FAIL")
            return -1
        elif error_code == ErrorCode.ERROR_VERIFY_NO_GNSSPOS:
            logging.error("ErrorCode.ERROR_VERIFY_NO_GNSSPOS")
            return -1
        elif error_code == ErrorCode.ERROR_NOT_STATIC:
            logging.warn("ErrorCode.ERROR_NOT_STATIC")
        else:
            logging.error("error error_code [%s]" % str(error_code))
        return 0

    def _start(self):
        request = collection_check_message_pb2.StaticAlignRequest(cmd=1)
        logging.info("static align start request: " + str(request))
        response = self.stub.StaticAlign(request)
        logging.info("static align start response: " + str(response))
        if response.code != ErrorCode.SUCCESS:
            return self._exception_handler(response.code)
        return 0
 
    def _check(self):
        request = collection_check_message_pb2.StaticAlignRequest(cmd=2)
        logging.info("static align check request: " + str(request))
        response = self.stub.StaticAlign(request)
        logging.info("static align check response: " + str(response))
        if response.code != ErrorCode.SUCCESS:
            ret = self._exception_handler(response.code)
            if ret != 0:
                return [-1, 0.0]
        return [0, response.progress]

    def _stop(self):
        request = collection_check_message_pb2.StaticAlignRequest(cmd=3)
        logging.info("static align stop request: " + str(request))
        response = self.stub.StaticAlign(request)
        logging.info("static align stop response: " + str(response))
        if response.code != ErrorCode.SUCCESS:
            logging.info("abnormal response: [%s]" % str(response))
            ret = self._exception_handler(response.code)
            if ret != 0:
                return [-1, 0.0]
        return [0, response.progress]

    def sync_start(self):
        ret = self._start()
        if ret != 0:
            logging.error("static align start failed")
            return -1
        while True:
            [ret, progress] = self._check()
            if ret != 0:
                logging.error("static align check failed")
                break
            logging.info("static align check progress: %f" % progress)
            if abs(progress - 1.0) < 1e-8:
                break
            time.sleep(self._conf["check_period"])
        if ret != 0:
            return -1
        [ret, _] = self._stop()
        if ret != 0:
            logging.error("static align stop failed")
            return -1
        return 0

if __name__ == "__main__":
    static_align = StaticAlign("127.0.0.1", "50100", "client.yaml")

File: modules/map/tools/map_datachecker/server/static_align.cc
/******************************************************************************
 * Copyright 2019 The Apollo Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *****************************************************************************/
#include "modules/map/tools/map_datachecker/server/static_align.h"

#include <vector>

namespace apollo {
namespace hdmap {

StaticAlign::StaticAlign(std::shared_ptr<JSonConf> sp_conf)
    : Alignment(sp_conf) {
  sp_conf_ = sp_conf;
  static_align_detect_method_ = StaticAlignDetectMethod::DYNAMIC_CENTROID;
  Reset();
}

void StaticAlign::Reset() {
  progress_ = 0.0;
  last_progress_ = 0.0;
  start_time_ = -1.0;
  end_time_ = -1.0;
  start_index_ = -1;
  end_index_ = -1;
  sp_bad_pose_info_ = std::make_shared<BadOrGoodPoseInfo>();
  sp_good_pose_info_ = std::make_shared<BadOrGoodPoseInfo>();
  dynamic_centroid_ = Centroid3D();
}

bool StaticAlign::IsStaticPose(const FramePose& pose) {
  if (dynamic_centroid_.count == 0) {
    return true;
  }
  double move_dist_x = pose.tx - dynamic_centroid_.center.x;
  double move_dist_y = pose.ty - dynamic_centroid_.center.y;
  double move_dist_z = pose.tz - dynamic_centroid_.center.z;
  double move_dist =
      std::sqrt(move_dist_x * move_dist_x + move_dist_y * move_dist_y +
                move_dist_z * move_dist_z);
  AINFO << "dist thresh: " << sp_conf_->static_align_dist_thresh
        << ", dist: " << move_dist;
  if (move_dist <= sp_conf_->static_align_dist_thresh) {
    return true;
  }
  return false;
}

void StaticAlign::UpdateDynamicCentroid(const FramePose& pose) {
  int count = dynamic_centroid_.count;
  if (count == 0) {
    dynamic_centroid_.start_time = pose.time_stamp;
  } else {
    dynamic_centroid_.end_time = pose.time_stamp;
  }
  AINFO << "cetroid start: " << dynamic_centroid_.start_time
        << ", end: " << dynamic_centroid_.end_time;

  double x = dynamic_centroid_.center.x * count + pose.tx;
  double y = dynamic_centroid_.center.y * count + pose.ty;
  double z = dynamic_centroid_.center.z * count + pose.tz;
  ++count;

  dynamic_centroid_.count = count;
  dynamic_centroid_.center.x = x / count;
  dynamic_centroid_.center.y = y / count;
  dynamic_centroid_.center.z = z / count;
}

double StaticAlign::GetCentroidTimeDuring() {
  if (dynamic_centroid_.start_time > 0 && dynamic_centroid_.end_time > 0) {
    return dynamic_centroid_.end_time - dynamic_centroid_.start_time;
  }
  return 0.0;
}

void StaticAlign::UpdateGoodPoseInfo(const FramePose& pose) {
  UpdateDynamicCentroid(pose);
}

double StaticAlign::StaticAlignDynamicCentroid(
    const std::vector<FramePose>& poses) {
  int start_index = TimeToIndex(poses, start_time_);
  AINFO << "start_index:" << start_index << ",pose size:" << poses.size();
  dynamic_centroid_ = Centroid3D();
  for (int i = start_index + 1; i < static_cast<int>(poses.size()); ++i) {
    if (!IsGoodPose(poses, i)) {
      AINFO << "not good pose";
      return_state_ = ErrorCode::ERROR_GNSS_SIGNAL_FAIL;
      return 0.0;
    }
    if (!IsStaticPose(poses[i])) {
      AINFO << "not static pose";
      return_state_ = ErrorCode::ERROR_NOT_STATIC;
      return 0.0;
    }
    UpdateGoodPoseInfo(poses[i]);
    return_state_ = ErrorCode::SUCCESS;
  }

  double progress = GetCentroidTimeDuring() / sp_conf_->static_align_duration;
  if (progress > 1.0) {
    progress = 1.0;
  }
  return progress;
}

double StaticAlign::StaticAlignRansac(const std::vector<FramePose>& poses) {
  // TODO(yuanyijun): implementation of selecting an center by RANSAC
  return 0.0;
}

double StaticAlign::GetStaticAlignProgress(
    const std::vector<FramePose>& poses) {
  double progress = 0.0;
  switch (static_align_detect_method_) {
    case StaticAlignDetectMethod::DYNAMIC_CENTROID:
      progress = StaticAlignDynamicCentroid(poses);
      break;
    case StaticAlignDetectMethod::RANSAC:
      progress = StaticAlignRansac(poses);
      break;
    default:
      break;
  }
  ClearGoodPoseInfo();
  return progress;
}

ErrorCode StaticAlign::Process(const std::vector<FramePose>& poses) {
  AINFO << "[StaticAlign::process] begin";
  size_t size = poses.size();
  if (size <= 1) {
    AINFO << "system has no pose, exit process";
    return_state_ = ErrorCode::ERROR_VERIFY_NO_GNSSPOS;
    return return_state_;
  }

  progress_ = GetStaticAlignProgress(poses);
  if (return_state_ != ErrorCode::SUCCESS) {
    AINFO << "get_static_align_progress error, progress 0.0";
    return return_state_;
  }

  AINFO << "[StaticAlign::process] end, progress:" << progress_;
  return_state_ = ErrorCode::SUCCESS;
  return return_state_;
}

}  // namespace hdmap
}  // namespace apollo
